syntax = "proto3";

package cargo;

service CargoService {
    rpc CreateCargo (CreateCargoRequest) returns (CreateCargoResponse);
    rpc GetUserCargos (GetUserCargosRequest) returns (GetUserCargosResponse);
    rpc UpdateCargo (UpdateCargoRequest) returns (UpdateCargoResponse);
    rpc GetCargo (GetCargoRequest) returns (GetCargoResponse);
}

message GetCargoRequest {
    string cargo_id = 1;
}

message GetCargoResponse {
    int32 code = 1;
    oneof GetCargoResponseOneOf {
        string message = 2;
        CargoData cargo_data = 3;
    }
}

message CreateCargoRequest {
    CreateCargoData creating_cargo_data = 1;
}


message CreateCargoResponse {
    int32 code = 1;
    oneof CreateCargoResponseOneOf {
        string message = 2;
        CargoData cargo_data = 3;
    }
}

message GetUserCargosRequest {
    int32 page = 1;
    string user_id = 2;
}

message GetUserCargosResponse {
    int32 code = 1;
    oneof GetUserCargosResponseOneOf {
        string message = 2;
        CargoDataArray arr = 3;
    }

}

message CargoDataArray {
    repeated CargoData cargo_data = 1;
}

message CreateCargoData {
    string title = 1;
    string type = 2;
    string description = 3;
    string creator_id = 4;
    int32 weight = 5;
}

message BriefDeliveryData {
    string id = 1;
    string state = 2; 
}

message UpdateCargoRequest {
    string cargo_id = 1;
    UpdateCargoData updating_cargo_data = 2;
}

message UpdateCargoResponse {
    int32 code = 1;
    oneof UpdateCargoResponseOneOf {
        string message = 2;
        CargoData cargo_data = 3;
    }
}

message UpdateCargoData {
    optional string title = 1;
    optional string type = 2;
    optional string description = 3;
    optional string creator_id = 4;
}

message CargoData {
    string id = 1;
    string title = 2;
    string type = 3;
    string description = 4;
    string creator_id = 5;
    optional BriefDeliveryData delivery_data = 6;
    int32 weight = 7;
}